Sming/SmingCore/HardwareSerial.cpp

```diff
@@ -162,129 +98,101 @@ void HardwareSerial::systemDebugOutput(bool enabled)
-							.uartNr = static_cast<uint8_t>(uart_nr)}};
-
-		if(m.HWSDelegate)
-			System.queueCallback(
-				[](os_param_t param) {
-					SerialParam ser = {.param = param};
-					auto& m = memberData[ser.uartNr];
-					if(m.HWSDelegate)
-						m.HWSDelegate(Serial, ser.receivedChar, ser.charCount);
-				},
-				ser.param);
-
+		SerialParam serialParam = {{.charCount = static_cast<uint16_t>(uart_rx_available(uart)),
+									.receivedChar = (uint8_t)uart_peek_last_char(uart),
+									.uartNr = static_cast<uint8_t>(uartNr)}};
+
+		System.queueCallback(
+			[](uint32_t param) {
+				SerialParam serialParam = {.param = param};
+				auto uart = uart_get_uart(serialParam.uartNr);
+				auto serial = reinterpret_cast<HardwareSerial*>(uart_get_callback_param(uart));
+				if(serial->HWSDelegate) {
+					serial->HWSDelegate(*serial, serialParam.receivedChar, serialParam.charCount);
+				}
 #if ENABLE_CMD_EXECUTOR
-		if(m.commandExecutor)
-			System.queueCallback(
-				[](os_param_t param) {
-					SerialParam ser = {.param = param};
-					auto& m = memberData[ser.uartNr];
-					if(m.commandExecutor)
-						m.commandExecutor->executorReceive(ser.receivedChar);
-				},
-				ser.param);
-
+				if(serial->commandExecutor) {
+					serial->commandExecutor->executorReceive(serialParam.receivedChar);
+				}
 #endif
+			},
+			serialParam.param);
 	}
 }
 
-bool HardwareSerial::setCallback(StreamDataReceivedDelegate reqDelegate)
+bool HardwareSerial::updateUartCallback()
 {
-	if(!uart_rx_enabled(uart)) {
+	if(uart == nullptr) {
 		return false;
 	}
 
-	uart->callback = callbackHandler;
-
-	memberData[uartNr].HWSDelegate = reqDelegate;
-	return true;
-}
-
-void HardwareSerial::resetCallback()
-{
-	memberData[uartNr].HWSDelegate = nullptr;
-
-	uart->callback = nullptr;
+#if ENABLE_CMD_EXECUTOR
+	if(HWSDelegate || transmitComplete || commandExecutor) {
+#else
+	if(HWSDelegate || transmitComplete) {
+#endif
+		setUartCallback(
+			[](uart_t* uart, uint32_t status) {
+				auto serial = reinterpret_cast<HardwareSerial*>(uart_get_callback_param(uart));
+				if(serial) {
+					serial->callbackHandler(status);
+				}
+			},
+			this);
+		return true;
+	} else {
+		setUartCallback(nullptr, nullptr);
+		return false;
+	}
 }
 
 void HardwareSerial::commandProcessing(bool reqEnable)
 {
 #if ENABLE_CMD_EXECUTOR
 	if(reqEnable) {
-		if(!memberData[uartNr].commandExecutor) {
-			memberData[uartNr].commandExecutor = new CommandExecutor(&Serial);
+		if(!commandExecutor) {
+			commandExecutor = new CommandExecutor(this);
 		}
 	} else {
-		delete memberData[uartNr].commandExecutor;
-		memberData[uartNr].commandExecutor = nullptr;
+		delete commandExecutor;
+		commandExecutor = nullptr;
 	}
+	updateUartCallback();
 #endif
 }
-
-int HardwareSerial::baudRate(void)
-{
-	// Null pointer on _uart is checked by SDK
-	return uart_get_baudrate(uart);
-}
-
-HardwareSerial::operator bool() const
-{
-	return uart != nullptr;
-}
-
-size_t HardwareSerial::indexOf(char c)
-{
```

Comment:
I want to fix the return type of this method, but that requires changing it in `Stream` as well. These are the only two classes affected in Sming, but it's potentially a BC. What say you?

Reply:
> return type of this method (should be int) ... What say you?

Good catch! Please make the changes and add in the commit message information about the BC. The impact should be relatively low.